src/software/ai/navigator/path_planner/hrvo/hrvo_agent.cpp

```diff
@@ -0,0 +1,653 @@
+#include "software/ai/navigator/path_planner/hrvo/hrvo_agent.h"
+
+HRVOAgent::HRVOAgent(RobotId robot_id, const RobotState &robot_state,
+                     const RobotPath &path, double radius, double max_speed,
+                     double max_accel, double max_radius_inflation)
+    : Agent(robot_id, robot_state, path, radius, max_speed, max_accel,
+            max_radius_inflation),
+      obstacle_factory(TbotsProto::RobotNavigationObstacleConfig()),
+      neighbours()
+{
+}
+
+void HRVOAgent::updatePrimitive(const TbotsProto::Primitive &new_primitive,
+                                const World &world, Duration time_step)
+{
+    RobotPath path;
+    static_obstacles.clear();
+    ball_obstacle = std::nullopt;
+    if (new_primitive.has_move())
+    {
+        const auto &motion_control = new_primitive.move().motion_control();
+        float speed_at_dest        = new_primitive.move().final_speed_m_per_s();
+        float new_max_speed        = new_primitive.move().max_speed_m_per_s();
+        this->max_speed            = new_max_speed;
+
+        // TODO (#2418): Update implementation of Primitive to support
+        // multiple path points and remove this check
+        CHECK(motion_control.path().points().size() >= 2)
+            << "Empty path: " << motion_control.path().points().size() << std::endl;
+        auto destination = motion_control.path().points().at(1);
+
+        // Max distance which the robot can travel in one time step + scaling
+        // TODO (#2370): This constant is calculated multiple    times.
+        double path_radius = (max_speed * time_step.toSeconds()) / 2;
+        auto path_points   = {PathPoint(
+            Point(destination.x_meters(), destination.y_meters()), speed_at_dest)};
+        path               = RobotPath(path_points, path_radius);
+
+        // Update static obstacles
+        std::set<TbotsProto::MotionConstraint> motion_constraints;
+        for (int constraint_int : motion_control.motion_constraints())
+        {
+            if (!TbotsProto::MotionConstraint_IsValid(constraint_int))
+            {
+                continue;
+            }
+
+            const auto constraint =
+                static_cast<TbotsProto::MotionConstraint>(constraint_int);
+            auto new_obstacles =
+                obstacle_factory.createFromMotionConstraint(constraint, world);
+            if (constraint == TbotsProto::MotionConstraint::HALF_METER_AROUND_BALL)
+            {
+                ball_obstacle = new_obstacles[0];
+            }
```

Comment:
we're assuming that the first obstacle if we have this constraint HALF_METER_AROUND_BALL is HALF_METER_AROUND_BALL which is too implementation-dependent

it's okay if this is in right now, but in the future let's track an issue to update this implementation

Reply:
My other open PR will generalize this to work with any obstacle and remove this assumption